python/ray/train/v2/_internal/execution/train_fn_utils.py

```diff
@@ -0,0 +1,89 @@
+import threading
+from typing import Any, Dict, Optional
+
+from ray.data import DataIterator
+from ray.train import Checkpoint
+from ray.train.v2._internal.execution.context import (
+    get_train_context as get_internal_train_context,
+)
+from ray.train.v2.api.context import TrainContext as PublicTrainContext
+
+
+class TrainFnUtils:
+    """Utility class providing an abstraction layer between user-facing APIs
+        and :class:`~ray.train.v2._internal.execution.context.TrainContext`.
+
+    It should be set before the users' training function is called, like training workers initialization.
+    This class can be patched if new user APIs behaviors is wanted.
+    """
+
+    def report(
+        self,
+        metrics: Dict[str, Any],
+        checkpoint: Optional[Checkpoint] = None,
+        checkpoint_dir_name: Optional[str] = None,
+    ) -> None:
+        """Upload checkpoint to remote storage and put a training result on the result queue.
+
+        Args:
+            metrics: The metrics to report.
+            checkpoint: The checkpoint to report.
+            checkpoint_dir_name: The name of the checkpoint dir
+                in this iteration. Note: If not set, the checkpoint will
+                be stored in the default storage path. If set, make sure
+                this value is unique for each iteration.
+        """
+        return get_internal_train_context().report(
+            metrics, checkpoint, checkpoint_dir_name
+        )
+
+    def get_synchronization_actor(self):
+        """Get the synchronization actor for collective operations among training workers.
+
```

Comment:
This one's not a user-facing API. Can we just use `get_internal_train_context` to access this?

And for now just exclude `ray.train.collective` methods from the scope of this `TrainFnUtils`. Put a todo to deal with that later.

We may need to create a shell `ray.train.collective.barrier` shell method that calls `train_fn_utils().barrier()`, which then routes to the actual implementation.